### ISSUE TITLE
error accessing constructor of `err` in the error serializer

### ISSUE THREAD

**Issue by AVVS** [pinojs/pino-std-serializers#31]
Recently encountered a weird error, here is the stack trace (node 12.18.1):

```
TypeError: Cannot read property 'name' of undefined
    at Object.errSerializer [as err] (/src/node_modules/pino-std-serializers/lib/err.js:45:31)
    at Pino.asJson (/src/node_modules/pino/lib/tools.js:118:50)
    at Pino.write (/src/node_modules/pino/lib/proto.js:160:28)
    at Pino.LOG [as info] (/src/node_modules/pino/lib/tools.js:55:21)
    at Server.audit (/src/lib/plugins/audit.js:158:15)
    at Server.emit (events.js:315:20)
```

https://github.com/pinojs/pino-std-serializers/blob/master/lib/err.js#L45

```
  _err.type = err.constructor.name
```

Incoming object is from restify `after` event:
Relevant piece of code where the log is triggered:

```
  function audit(req, res, route, err) {
    let latency = res.get('Response-Time');

    if (typeof latency !== 'number') {
      latency = Date.now() - req._time;
    }

    const obj = {
      remoteAddress: req.connection.remoteAddress,
      remotePort: req.connection.remotePort,
      user: req.user && req.user.id,
      req_id: req.getId(),
      req,
      res,
      err,
      latency,
      secure: req.secure,
      _audit: true,
      event: opts.event
    };

    if (printLog) {
      switch (opts.event) {
        case 'after':
          log.info(obj, 'handled: %d', res.statusCode); // <---------- error was triggered here
          break;

        case 'pre':
          log.info(obj, 'pre');
          break;

        case 'routed':
          log.info(obj, 'routed');
          break;

        default:
          throw new Error(`Unexpected audit event: ${opts.event}`);
      }
    }

    if (server !== undefined) {
      server.emit('audit', obj);
    }

    return true;
  }

  return audit;
}
```

I'm not exactly sure how it is possible for `err.constructor` to be undefined and still be an `instanceof Error`, but the error did happen.

Should I create a PR that introduces a check on the constructor property? I'm not sure how detailed the check should be as to not affect the performance too much

ie (truthy/instanceof) or smth else? or is the reason for error entirely in my domain and I should dig deeper to understand the reason for missing constructor?

**Comment by jsumners** [comment]
A minimal reproduction would be the first step in determining the answer to any of your questions.

**Comment by AVVS** [comment]
@jsumners I'd be happy to make one except I don't have a slightest idea how this happened. I've got that stacktrace in the server logs, which, then crashed the server. the most weird part here is that the server process typically runs for months without crashes and handles hundreds of thousands of requests per day, so it would've been an extremely rare occurrence

**Comment by mcollina** [comment]
I think this would be easy to fix. Would you like to send a Pull Request to address this issue? Remember to add unit tests.

(Note that you have something weird happening in your codebase).

**Comment by AVVS** [comment]
I’d absolutely submit a PR

In terms of having an issue in a codebase somewhere - definitely agree, but how do I go about finding it? This did happen once, skimmed through all log calls and nothing seem to be doing any weird reassignment of constructor property - any suggestions that might help except putting a log with a check on err.constructor and reporting this somewhere with a stack if it ever happens again?

**Comment by mcollina** [comment]
I'd put a custom serializer for err in that log where an error with no constructor is passed in.

**Comment by voxpelli** [comment]
Seems fixed by #47?
